### ISSUE TITLE
After running in a cluster with ~200 pods for over 2 days, `linkerd dashboard` no longer works

### ISSUE THREAD

**Issue by hawkw** [linkerd/linkerd2#1406]
I noticed this issue while working on the verification for the changes in #1358.

After running a `linkerd-lifecycle` test with over 200 pods (scaled with `bin/scale 5 5`) for over two days, I can no longer open the web dashboard or Grafana using the `linkerd dashboard` command:
```
➜ bin/linkerd dashboard -l linkerd-lifecycle
Linkerd is not running in the "linkerd-lifecycle" namespace
Install with: linkerd install --linkerd-namespace linkerd-lifecycle | kubectl apply -f -
```

```
➜ bin/linkerd dashboard --show grafana -l linkerd-lifecycle
Linkerd is not running in the "linkerd-lifecycle" namespace
Install with: linkerd install --linkerd-namespace linkerd-lifecycle | kubectl apply -f -
```

I've retried these commands repeatedly, and they always fail with the same message --- it's not a transient issue.

However, `kubectl` indicates that the requisite pods _are_ running:
```
➜ kubectl get po -n linkerd-lifecycle
NAME                          READY     STATUS    RESTARTS   AGE
controller-544b9bcf6d-6plbm   5/5       Running   0          2d
grafana-677779484d-7v59f      2/2       Running   0          2d
prometheus-757fdd4b44-vfd9l   2/2       Running   0          2d
web-78d967d487-jt5zf          2/2       Running   0          2d
```

The `linkerd check` output indicates that querying the Linkerd API is timing out: 
```
➜ bin/linkerd check -l linkerd-lifecycle
kubernetes-api: can initialize the client..................................[ok]
kubernetes-api: can query the Kubernetes API...............................[ok]
kubernetes-api: is running the minimum Kubernetes API version..............[ok]
linkerd-api: can query the Linkerd API.....................................[ERROR] -- Post https://147.75.70.69:6443/api/v1/namespaces/linkerd-lifecycle/services/http:api:http/proxy/api/v1/SelfCheck: context deadline exceeded
linkerd-version: cli is up-to-date.........................................[FAIL]  -- is running version git-c0351933 but the latest version is v18.7.3
linkerd-version: control plane is up-to-date...............................[FAIL]  -- is running version git-c0351933 but the latest version is v18.7.3
```
I suspect that's waht happens when I run the `dashboard` command as well?

**Comment by siggy** [comment]
A few more notes...

Requests from Public API -> Prometheus appear to fail:
```bash
$ kubectl -n linkerd-lifecycle logs -f $(kubectl -n linkerd-lifecycle get po --selector=linkerd.io/control-plane-component=controller -o jsonpath='{.items[*].metadata.name}') public-api
...
time="2018-08-06T20:25:57Z" level=error msg="Query(max(process_start_time_seconds{}) by (pod, namespace)) failed with: bad_response: bad response code 500"
time="2018-08-06T20:26:21Z" level=error msg="Query(max(process_start_time_seconds{}) by (pod, namespace)) failed with: bad_response: bad response code 500"
```

Requests from the linkerd-proxy in the Public API pod to Prometheus succeed:
```bash
$ kubectl -n linkerd-lifecycle exec -it $(kubectl -n linkerd-lifecycle get po --selector=linkerd.io/control-plane-component=controller -o jsonpath='{.items[*].metadata.name}') /bin/bash -c linkerd-proxy
...
$ curl -s -o /dev/null -w "%{http_code}" "http://prometheus.linkerd-lifecycle.svc.cluster.local:9090/api/v1/query?query=max(process_start_time_seconds%7B%7D)%20by%20(pod%2C%20namespace)"
200
```

Prometheus reports itself as healthy:

```bash
$ kubectl -n linkerd-lifecycle port-forward $(kubectl -n linkerd-lifecycle get po --selector=linkerd.io/control-plane-component=prometheus -o jsonpath='{.items[*].metadata.name}') 9090:9090
...

$ curl localhost:9090/-/healthy
Prometheus is Healthy.
```

The Linkerd Proxy running in the Prometheus pod is logging a lot of errors:
```bash
$ kubectl -n linkerd-lifecycle logs -f $(kubectl -n linkerd-lifecycle get po --selector=linkerd.io/control-plane-component=prometheus -o jsonpath='{.items[*].metadata.name}') linkerd-proxy

ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.233.66.155:56212} linkerd2_proxy turning operation timed out after 10s into 500
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.233.66.155:53652} linkerd2_proxy turning an error occurred trying to connect: No route to host (os error 113) into 500
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.233.66.155:60924} linkerd2_proxy turning an error occurred trying to connect: Connection refused (os error 111) into 500
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.233.66.155:40374} linkerd2_proxy turning an error occurred trying to connect: No route to host (os error 113) into 500
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.233.66.155:41472} linkerd2_proxy turning operation timed out after 10s into 500
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.233.66.155:39654} linkerd2_proxy turning operation timed out after 10s into 500
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.233.66.155:58498} linkerd2_proxy turning operation timed out after 10s into 500
```

Full metrics dump from Linkerd proxy in Prometheus pod:

```bash
kubectl -n linkerd-lifecycle exec -it $(kubectl -n linkerd-lifecycle get po --selector=linkerd.io/control-plane-component=prometheus -o jsonpath='{.items[*].metadata.name}') curl localhost:4191/metrics -c linkerd-proxy
```
generates:
https://gist.github.com/siggy/3d6eba2743a0f35040a86ac552ce5ddd

**Comment by hawkw** [comment]
After rolling the Prometheus and controller pods (accidentally), the cluster has been up for about three more days and the issue hasn't reoccurred yet.

**Comment by chrisgoffinet** [comment]
Next time this happens, it might be worthwhile to dump out how many file descriptors are open, wondering if a leak.

**Comment by hawkw** [comment]
@chrisgoffinet I don't _think_ this is caused by a FD leak, as the Grafana and Web pods do respond to requests from inside the cluster.

I did some investigation, and it appears that this issue is caused by the Prometheus health check (a very simple Prometheus query) timing out when run by the controller pod, so it considers Prometheus to be 'down'. However, when I manually connected to Prometheus and ran the same query, it ran in under 10ms most of the time --- I only saw elevated latencies very sporadically.
